crates/sp-domains-fraud-proof/src/verification.rs

```diff
@@ -317,3 +321,142 @@ where
+                })
+            }
+            BundleValidity::Valid(valid_bundle_hash) => {
+                Err(VerificationError::UnexpectedValidBundleEntry {
+                    bundle_index: invalid_bundle_fraud_proof.bundle_index,
+                    bundle_hash: valid_bundle_hash.clone().into(),
+                })
+            }
+        };
+    }
+
+    Ok(targeted_invalid_bundle_entry.clone())
+}
+
+pub fn verify_invalid_bundles_fraud_proof<CBlock, DomainHeader, Balance>(
+    bad_receipt: ExecutionReceipt<
+        NumberFor<CBlock>,
+        CBlock::Hash,
+        HeaderNumberFor<DomainHeader>,
+        HeaderHashFor<DomainHeader>,
+        Balance,
+    >,
+    invalid_bundles_fraud_proof: &InvalidBundlesFraudProof<HeaderHashFor<DomainHeader>>,
+) -> Result<(), VerificationError>
+where
+    CBlock: BlockT,
+    CBlock::Hash: Into<H256>,
+    DomainHeader: HeaderT,
+    DomainHeader::Hash: Into<H256> + From<H256> + PartialEq + Clone,
+{
+    let invalid_bundle_entry = check_expected_bundle_entry::<
+        CBlock,
+        <DomainHeader as HeaderT>::Number,
+        <DomainHeader as HeaderT>::Hash,
+        Balance,
+        <DomainHeader as HeaderT>::Hash,
+    >(&bad_receipt, invalid_bundles_fraud_proof)?;
+
+    let extrinsic = StorageProofVerifier::<BlakeTwo256>::get_decoded_value(
+        &invalid_bundle_entry.extrinsics_root.into(),
+        StorageProof::new(
+            invalid_bundles_fraud_proof
+                .extrinsic_inclusion_proof
+                .clone()
+                .drain(..),
+        ),
+        StorageKey(
+            invalid_bundles_fraud_proof
+                .mismatched_extrinsic_index
+                .to_be_bytes()
+                .to_vec(),
+        ),
+    )
+    .map_err(|_e| VerificationError::InvalidProof)?;
+
+    match invalid_bundles_fraud_proof.proof_data {
+        ProofDataPerInvalidBundleType::OutOfRangeTx => {
+            let is_tx_in_range = fraud_proof_runtime_interface::get_fraud_proof_verification_info(
+                H256::from_slice(bad_receipt.consensus_block_hash.as_ref()),
+                FraudProofVerificationInfoRequest::TxRangeCheck {
+                    domain_id: invalid_bundles_fraud_proof.domain_id,
+                    bundle_index: invalid_bundles_fraud_proof.bundle_index,
+                    opaque_extrinsic: extrinsic,
+                },
+            )
+            // These are two different kind of errors as the first represent that we got None from the host function
+            // while second represents that we got the FraudProofVerificationInfoResponse but was none.
+            .ok_or(VerificationError::FailedToGetResponseFromTxRangeHostFn)? // Not possible to return any more info as we do not receive actual error here
+            .into_tx_range_check()
+            .ok_or(VerificationError::ReceivedInvalidInfoFromTxRangeHostFn)?; // Not possible to return any more info as we do not receive actual error here
```

Comment:
These comments seem annoying and unnecessary IMHO.

Reply:
Yeah, plz make it to one TODO, and "These are two different kind of err..." also seems unnecessary to me, because the name already explains.